### PR TITLE
Make nWalletDBUpdated atomic to avoid a potential race.

### PR DESCRIPTION
**Pull request by pstratem**
Title

### CODE REVIEW

**Comment by luke-jr** [comment]
utACK

**Comment by TheBlueMatt** [comment]
If you're gonna touch this, might as well fix it - will it interact negatively with multiwallet support as written? Also, should be easy to make it only exist in walletdb.cpp, I'd think?

**Comment by laanwj** [comment]
> will it interact negatively with multiwallet support as written? 

Right - this should probably exist on the CWalletDB object instead of globally, otherwise multiple wallets will react on each other's changes.
This is not entirely trivial to fix because the 'wallet flush' thread (which IIRC uses this) is decidedly single-wallet. Maybe one of the multiwallet pulls addresses this though.

**Comment by jonasschnelli** [comment]
Agree with @laanwj.
Also, #8776 would be a step towards a better – multiwallet safe – wallet flush.

**Comment by luke-jr** [comment]
multiwallet makes the flush thread go over all the wallets and flush them all when this is updated. AFAIK extraneous flushing is mostly harmless...?

**Comment by laanwj** [comment]
> multiwallet makes the flush thread go over all the wallets and flush them all when this is updated. AFAIK extraneous flushing is mostly harmless...?

It's would be a performance sink. Remember the thread is named badly. Our 'flushing' isn't really flushing, it is a full fledged database consolidation.

**Comment by TheBlueMatt** [comment]
This should probably also move to CScheduler to avoid having Yet Another Thread when we dont need it.

**Comment by gmaxwell** [comment]
The db flush here doesn't really flush a wallet at all, it flushes the whole database environment, so I don't believe it can be made per-wallet.

**Comment by pstratem** [comment]
Indeed this cannot be made per wallet, this is flushing the entire database environment.

**Comment by jonasschnelli** [comment]
Right now we are flushing the DB environment (https://docs.oracle.com/cd/E17276_01/html/api_reference/C/envclose.html) with `DBEnv->CloseDb(filename)`.
Is there a reason why we wouldn't want to use `DB->Sync()` (https://docs.oracle.com/cd/E17276_01/html/api_reference/C/dbsync.html)?

**Comment by laanwj** [comment]
> Is there a reason why we wouldn't want to use DB->Sync() 
> (https://docs.oracle.com/cd/E17276_01/html/api_reference/C/dbsync.html)?

The wallet code is set up to *flush* after every (important) write immediately, that's not what the thread is for.

What the flush thread does is not really flushing. The method you propose would be a less hassle solution to do flushing, but what the thread wants to do is consolidate the database files: When writing, BerkeleyDB writes to log files and merges these into the main database only incidentally. The whole circuitous exercise in the thread is to force this to happen, so that the wallet is one file.

(can we **please** rename this thread and add a comment somewhere, everyone reading this code is confused about this)

**Comment by laanwj** [comment]
> The db flush here doesn't really flush a wallet at all, it flushes the whole database environment, so I don't believe it can be made per-wallet.

Doesn't every wallet get its own database environment?

**Comment by gmaxwell** [comment]
No-- well every database environment needs it's own database directory, cache overheads, etc. So I'm guessing that wouldn't be a great move.

File: src/wallet/walletdb.h [hunk 1]
```diff
@@ -11,6 +11,7 @@
 #include "wallet/db.h"
 #include "key.h"
 
+#include <atomic>
```
Comment (TheBlueMatt) [inline src/wallet/walletdb.h]: nit: move this to the .cpp.

Comment (pstratem) [inline src/wallet/walletdb.h]: nit picked

Comment (TheBlueMatt) [inline src/wallet/walletdb.h]: Not it wasnt?

**Comment by TheBlueMatt** [comment]
Can this get an 0.14 tag, since its a simple bugfix?

**Comment by TheBlueMatt** [comment]
utACK d63ff6265b0c6ae30efcbb9120d4db419606198a

**Comment by sipa** [comment]
utACK d63ff6265b0c6ae30efcbb9120d4db419606198a

**Pull request merged**